### PR TITLE
Ensure serialized file is little endian

### PR DESCRIPTION
**Pull request by kddnewton**
So that it's portable to other machines

### CODE REVIEW

File: templates/src/serialize.c.erb [hunk 1]
```diff
@@ -47,6 +47,31 @@ pm_serialize_string(pm_parser_t *parser, pm_string_t *string, pm_buffer_t *buffe
...
+static void
+pm_serialize_32(char *address, uint32_t value) {
+    static bool little_endian = false;
+    static bool endian_check = false;
+
+    if (!endian_check) {
+        uint32_t test = 0x1;
+        little_endian = *((uint8_t *) &test) == 0x1;
+        endian_check = true;
+    }
```
[the remark this answers is not in the document]
Comment (eregon) [reply templates/src/serialize.c.erb:63]: CRuby seems to use `WORDS_BIGENDIAN` which seems defined as:
```c
#ifdef AC_APPLE_UNIVERSAL_BUILD
# undef WORDS_BIGENDIAN
# ifdef __BIG_ENDIAN__
#  define WORDS_BIGENDIAN
# endif
#endif
```

File: templates/java/org/prism/Loader.java.erb [hunk 2]
```diff
@@ -54,6 +54,8 @@ public class Loader {
 
             if (constant == null) {
                 int offset = bufferOffset + index * 8;
+
+                // Read an integer in forced little endian.
```
Comment (eregon) [inline templates/java/org/prism/Loader.java.erb:58]: I would remove this, that should be clear from the `this.buffer = ByteBuffer.wrap(serialized).order(ByteOrder.LITTLE_ENDIAN);` line below.

**Review by eregon — COMMENTED** [review]
Looks good.
One improvement which I think is important is to use the preprocessor like `#ifndef WORDS_BIGENDIAN` instead of a runtime check like `if (little_endian) {`.

File: templates/src/serialize.c.erb (hunk 1 above)
[the remark this answers is not in the document]
Comment (eregon) [reply templates/src/serialize.c.erb:63]: From https://github.com/ruby/ruby/blob/e6b2cd15e34cbeaf07d8ced2db89d02272e9d5af/include/ruby/internal/config.h#L85-L90

**Review by seven1m — COMMENTED** [review]
Nice!

File: templates/src/serialize.c.erb [hunk 3]
```diff
@@ -47,6 +47,21 @@ pm_serialize_string(pm_parser_t *parser, pm_string_t *string, pm_buffer_t *buffe
     }
 }
 
+/**
+ * Serialize a 32-bit integer to the given address always in little-endian.
+ */
+static void
+pm_serialize_32(char *address, uint32_t value) {
+    if (PRISM_WORDS_BIGENDIAN) {
```
Comment (eregon) [inline templates/src/serialize.c.erb:55]:
```suggestion
#ifdef PRISM_WORDS_BIGENDIAN
```
So the C compiler doesn't even see the code for the wrong endian.

Comment (kddnewton) [reply templates/src/serialize.c.erb]: 👍

**Comment by eregon** [comment]
FWIW it seems ruby/ruby has RB_IO_BUFFER_HOST_ENDIAN, so we could likely use similar logic as that:
https://bugs.ruby-lang.org/issues/20431#note-5

**Pull request merged**